File: Sources/DebridStreamer/Services/AI/OpenAIProvider.swift
import Foundation

struct OpenAIProvider: AIAssistantProvider {
    let kind: AIProviderKind = .openAI

    private let apiKey: String
    private let model: String
    private let session: URLSession

    init(apiKey: String, model: String = "gpt-4o-mini", session: URLSession = .shared) {
        self.apiKey = apiKey
        self.model = model
        self.session = session
    }

    func recommend(prompt: String, candidateTitles: [String], maxResults: Int) async throws -> AIProviderRecommendationResult {
        let trimmedKey = apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedKey.isEmpty else {
            throw AIAssistantProviderError.missingAPIKey
        }

        guard let url = URL(string: "https://api.openai.com/v1/chat/completions") else {
            throw URLError(.badURL)
        }

        let envelope = AIAssistantJSONParser.promptEnvelope(
            userPrompt: prompt,
            candidateTitles: candidateTitles,
            maxResults: maxResults
        )

        let payload = OpenAIChatRequest(
            model: model,
            messages: [
                .init(role: "system", content: "You produce concise recommendations in JSON."),
                .init(role: "user", content: envelope)
            ],
            temperature: 0.4
        )

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 45
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(trimmedKey)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONEncoder().encode(payload)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw AIAssistantProviderError.invalidResponse
        }
        guard (200...299).contains(http.statusCode) else {
            let errorText = String(data: data, encoding: .utf8) ?? "OpenAI error"
            throw AIAssistantProviderError.apiError(errorText)
        }

        let decoded: OpenAIChatResponse
        do {
            decoded = try JSONDecoder().decode(OpenAIChatResponse.self, from: data)
        } catch {
            throw AIAssistantProviderError.invalidResponse
        }
        guard let content = decoded.choices.first?.message.content else {
            throw AIAssistantProviderError.invalidResponse
        }
        let recommendations = AIAssistantJSONParser.parseRecommendations(from: content, maxResults: maxResults)
        let usage = AIUsageMetrics(
            inputTokens: decoded.usage?.promptTokens,
            outputTokens: decoded.usage?.completionTokens,
            totalTokens: decoded.usage?.totalTokens,
            estimatedCostUSD: AIUsageCostEstimator.estimateUSD(
                model: decoded.model ?? model,
                inputTokens: decoded.usage?.promptTokens,
                outputTokens: decoded.usage?.completionTokens,
                totalTokens: decoded.usage?.totalTokens
            )
        )
        return AIProviderRecommendationResult(
            model: decoded.model ?? model,
            recommendations: recommendations,
            rawText: content,
            usage: usage
        )
    }

    /// Single-shot completion for the NL→filter mood discovery: sends `prompt`
    /// verbatim and returns the raw text (no recommendation envelope).
    func complete(prompt: String) async throws -> String {
        let trimmedKey = apiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmedKey.isEmpty else { throw AIAssistantProviderError.missingAPIKey }
        guard let url = URL(string: "https://api.openai.com/v1/chat/completions") else {
            throw URLError(.badURL)
        }

        let payload = OpenAIChatRequest(
            model: model,
            messages: [
                .init(role: "system", content: "You translate requests into JSON. Reply with JSON only."),
                .init(role: "user", content: prompt)
            ],
            temperature: 0.2
        )

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 45
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(trimmedKey)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONEncoder().encode(payload)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw AIAssistantProviderError.invalidResponse
        }
        guard (200...299).contains(http.statusCode) else {
            throw AIAssistantProviderError.apiError(String(data: data, encoding: .utf8) ?? "OpenAI error")
        }
        let decoded: OpenAIChatResponse
        do {
            decoded = try JSONDecoder().decode(OpenAIChatResponse.self, from: data)
        } catch {
            throw AIAssistantProviderError.invalidResponse
        }
        guard let content = decoded.choices.first?.message.content else {
            throw AIAssistantProviderError.invalidResponse
        }
        return content
    }
}

private struct OpenAIChatRequest: Encodable {
    struct Message: Encodable {
        let role: String
        let content: String
    }

    let model: String
    let messages: [Message]
    let temperature: Double
}

private struct OpenAIChatResponse: Decodable {
    struct Usage: Decodable {
        let promptTokens: Int?
        let completionTokens: Int?
        let totalTokens: Int?

        enum CodingKeys: String, CodingKey {
            case promptTokens = "prompt_tokens"
            case completionTokens = "completion_tokens"
            case totalTokens = "total_tokens"
        }
    }

    struct Choice: Decodable {
        struct Message: Decodable {
            let content: String
        }
        let message: Message
    }

    let model: String?
    let choices: [Choice]
    let usage: Usage?
}

File: web/src/services/ai/OpenAIProvider.ts
// Port of Sources/DebridStreamer/Services/AI/OpenAIProvider.swift.
//
// A fetch-based OpenAI Chat Completions provider. Mirrors the Swift struct: the
// same endpoint, Bearer auth, the system+user message pair, `temperature: 0.4`,
// the first choice's message content as the response text, the prompt/completion
// token usage decoding, and the cost estimate. The `fetch` implementation is
// injectable so tests stub the network.

import type {
  AIProviderKind,
  AIProviderRecommendationResult,
  AIUsageMetrics,
} from "./models";
import {
  type AIAnalyzeTitleInput,
  AIAssistantJSONParser,
  AIAssistantProviderError,
  type AIProviderAnalysisResult,
  AIUsageCostEstimator,
  type AIAssistantProvider,
  type FetchImpl,
  parsePersonalizedAnalysis,
  personalizedAnalysisPrompt,
  resolveFetch,
} from "./types";

const DEFAULT_BASE_URL = "https://api.openai.com/v1";

/** Config for pointing this provider at an OpenAI-compatible host (Gemini,
 * Groq, OpenRouter, Mistral, DeepSeek, xAI). Omit for stock OpenAI. */
export interface OpenAICompatConfig {
  /** Base URL WITHOUT a trailing `/chat/completions` (e.g. https://api.groq.com/openai/v1). */
  baseURL?: string;
  /** The kind this instance reports (drives usage records + display). */
  kind?: AIProviderKind;
  /** Human-facing name used in error messages ("Groq error"). */
  label?: string;
  /** Default model when the user hasn't chosen one. */
  defaultModel?: string;
}

// MARK: - Request shape (mirrors OpenAIChatRequest)

interface OpenAIChatRequestBody {
  model: string;
  messages: { role: string; content: string }[];
  temperature: number;
}

// MARK: - Raw response (mirrors OpenAIChatResponse, snake_case usage fields)

interface RawOpenAIUsage {
  prompt_tokens?: number | null;
  completion_tokens?: number | null;
  total_tokens?: number | null;
}

interface RawOpenAIChoice {
  message: { content: string };
}

interface RawOpenAIChatResponse {
  model?: string | null;
  choices: RawOpenAIChoice[];
  usage?: RawOpenAIUsage | null;
}

export class OpenAIProvider implements AIAssistantProvider {
  readonly kind: AIProviderKind;

  private readonly apiKey: string;
  private readonly model: string;
  private readonly fetchImpl: FetchImpl;
  private readonly chatURL: string;
  private readonly label: string;

  constructor(
    apiKey: string,
    model?: string,
    fetchImpl?: FetchImpl,
    config?: OpenAICompatConfig,
  ) {
    this.apiKey = apiKey;
    const chosen = model?.trim();
    this.model =
      chosen != null && chosen.length > 0
        ? chosen
        : config?.defaultModel ?? "gpt-4o-mini";
    this.fetchImpl = resolveFetch(fetchImpl);
    this.chatURL = `${config?.baseURL ?? DEFAULT_BASE_URL}/chat/completions`;
    this.kind = config?.kind ?? "openai";
    this.label = config?.label ?? "OpenAI";
  }

  async recommend(
    prompt: string,
    candidateTitles: string[],
    maxResults: number,
  ): Promise<AIProviderRecommendationResult> {
    const trimmedKey = this.apiKey.trim();
    if (trimmedKey.length === 0) {
      throw AIAssistantProviderError.missingAPIKey();
    }

    const envelope = AIAssistantJSONParser.promptEnvelope(
      prompt,
      candidateTitles,
      maxResults,
    );

    const payload: OpenAIChatRequestBody = {
      model: this.model,
      messages: [
        {
          role: "system",
          content: "You produce concise recommendations in JSON.",
        },
        { role: "user", content: envelope },
      ],
      temperature: 0.4,
    };

    const response = await this.fetchImpl(this.chatURL, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${trimmedKey}`,
      },
      body: JSON.stringify(payload),
    });

    if (!(response.status >= 200 && response.status <= 299)) {
      const errorText = (await response.text().catch(() => "")) || `${this.label} error`;
      throw AIAssistantProviderError.apiError(errorText);
    }

    const decoded = JSON.parse(await response.text()) as RawOpenAIChatResponse;
    const content = decoded.choices?.[0]?.message?.content;
    if (content == null) {
      throw AIAssistantProviderError.invalidResponse();
    }

    const recommendations = AIAssistantJSONParser.parseRecommendations(
      content,
      maxResults,
    );

    const resolvedModel = decoded.model ?? this.model;
    const inputTokens = decoded.usage?.prompt_tokens ?? null;
    const outputTokens = decoded.usage?.completion_tokens ?? null;
    const totalTokens = decoded.usage?.total_tokens ?? null;
    const usage: AIUsageMetrics = {
      inputTokens,
      outputTokens,
      totalTokens,
      estimatedCostUSD: AIUsageCostEstimator.estimateUSD(
        resolvedModel,
        inputTokens,
        outputTokens,
        totalTokens,
      ),
    };

    return {
      model: resolvedModel,
      recommendations,
      rawText: content,
      usage,
    };
  }

  async analyzeTitle(
    input: AIAnalyzeTitleInput,
  ): Promise<AIProviderAnalysisResult> {
    const trimmedKey = this.apiKey.trim();
    if (trimmedKey.length === 0) {
      throw AIAssistantProviderError.missingAPIKey();
    }

    const prompt = personalizedAnalysisPrompt(input);

    const payload: OpenAIChatRequestBody = {
      model: this.model,
      messages: [
        {
          role: "system",
          content:
            "You produce a single, strict JSON object analyzing a title for the user.",
        },
        { role: "user", content: prompt },
      ],
      temperature: 0.4,
    };

    const response = await this.fetchImpl(this.chatURL, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${trimmedKey}`,
      },
      body: JSON.stringify(payload),
    });

    if (!(response.status >= 200 && response.status <= 299)) {
      const errorText = (await response.text().catch(() => "")) || `${this.label} error`;
      throw AIAssistantProviderError.apiError(errorText);
    }

    const decoded = JSON.parse(await response.text()) as RawOpenAIChatResponse;
    const content = decoded.choices?.[0]?.message?.content;
    if (content == null) {
      throw AIAssistantProviderError.invalidResponse();
    }

    const analysis = parsePersonalizedAnalysis(content);

    const resolvedModel = decoded.model ?? this.model;
    const inputTokens = decoded.usage?.prompt_tokens ?? null;
    const outputTokens = decoded.usage?.completion_tokens ?? null;
    const totalTokens = decoded.usage?.total_tokens ?? null;

    return {
      model: resolvedModel,
      analysis,
      rawText: content,
      usage: {
        inputTokens,
        outputTokens,
        totalTokens,
        estimatedCostUSD: AIUsageCostEstimator.estimateUSD(
          resolvedModel,
          inputTokens,
          outputTokens,
          totalTokens,
        ),
      },
    };
  }
}
